File: app/src/debug/java/com/kaciula/archiman/util/log/LumberYard.java
package com.kaciula.archiman.util.log;

import android.annotation.SuppressLint;
import android.content.Context;
import android.support.annotation.NonNull;
import com.kaciula.archiman.R;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Deque;
import java.util.List;
import java.util.Locale;
import timber.log.Timber;

public class LumberYard {

  private static final boolean HAS_TIMBER;

  static {
    boolean hasDependency;

    try {
      Class.forName("timber.log.Timber");
      hasDependency = true;
    } catch (ClassNotFoundException exception) {
      hasDependency = false;
    }

    HAS_TIMBER = hasDependency;
  }

  private static final int BUFFER_SIZE = 200;

  private final DateFormat filenameDate = new SimpleDateFormat("yyyy-MM-dd hhmm a", Locale.US);
  private final DateFormat logDatePattern = new SimpleDateFormat("MM-dd hh:mm:ss.S", Locale.US);

  private static final String LOG_FILE_END = ".log";

  @SuppressLint("StaticFieldLeak") private static LumberYard sInstance;

  private final Context context;

  private final Deque<LogEntry> entries = new ArrayDeque<>(BUFFER_SIZE + 1);

  private OnLogListener onLogListener;

  public LumberYard(@NonNull Context context) {
    if (!HAS_TIMBER) {
      throw new RuntimeException("Timber dependency is not found");
    }
    this.context = context.getApplicationContext();
  }

  public static LumberYard getInstance(Context context) {
    if (sInstance == null) {
      sInstance = new LumberYard(context);
    }

    return sInstance;
  }

  public Timber.Tree tree() {
    return new Timber.DebugTree() {
      @Override
      protected void log(int priority, String tag, String message, Throwable t) {
        addEntry(new LogEntry(priority, tag, message,
            logDatePattern.format(Calendar.getInstance().getTime())));
      }
    };
  }

  public void setOnLogListener(OnLogListener onLogListener) {
    this.onLogListener = onLogListener;
  }

  private synchronized void addEntry(LogEntry entry) {
    entries.addLast(entry);

    if (entries.size() > BUFFER_SIZE) {
      entries.removeFirst();
    }

    onLog(entry);
  }

  public List<LogEntry> bufferedLogs() {
    return new ArrayList<>(entries);
  }

  /**
   * Save the current logs to disk.
   */
  public void save(OnSaveLogListener listener) {
    File dir = getLogDir();

    if (dir == null) {
      listener.onError(context.getString(R.string.all_save_logs_error));
      return;
    }

    FileWriter fileWriter = null;

    try {
      File output = new File(dir, getLogFileName());
      fileWriter = new FileWriter(output, true);

      List<LogEntry> entries = bufferedLogs();
      for (LogEntry entry : entries) {
        fileWriter.write(entry.prettyPrint() + "\n");
      }

      listener.onSave(output);

    } catch (IOException exception) {
      listener.onError(exception.getMessage());
      Timber.w(exception, exception.getMessage());
    } finally {
      if (fileWriter != null) {
        try {
          fileWriter.close();
        } catch (IOException exception) {
          listener.onError(exception.getMessage());
          Timber.w(exception, exception.getMessage());
        }
      }
    }
  }

  public void cleanUp() {
    File dir = getLogDir();
    if (dir != null) {
      File[] files = dir.listFiles();
      for (File file : files) {
        if (file.getName().endsWith(LOG_FILE_END)) {
          file.delete();
        }
      }
    }
  }

  private File getLogDir() {
    File file = new File(context.getFilesDir(), "logs");
    if (!file.exists()) {
      file.mkdirs();
    }
    return file;
  }

  private void onLog(LogEntry entry) {
    if (onLogListener != null) {
      onLogListener.onLog(entry);
    }
  }

  private String getLogFileName() {
    String pattern = "%s%s";
    String currentDate = filenameDate.format(Calendar.getInstance().getTime());

    return String.format(pattern, currentDate, LOG_FILE_END);
  }

  public interface OnSaveLogListener {
    void onSave(File file);

    void onError(String message);
  }


  public interface OnLogListener {
    void onLog(LogEntry logEntry);
  }
}

File: app/src/debug/java/com/kaciula/archiman/infrastructure/log/LumberYard.kt
package com.kaciula.archiman.infrastructure.log

import android.annotation.SuppressLint
import android.content.Context
import com.kaciula.archiman.R
import timber.log.Timber
import java.io.File
import java.io.FileWriter
import java.io.IOException
import java.text.SimpleDateFormat
import java.util.*

class LumberYard(context: Context) {

    private val filenameDate = SimpleDateFormat("yyyy-MM-dd hhmm a", Locale.US)
    private val logDatePattern = SimpleDateFormat("MM-dd hh:mm:ss.S", Locale.US)

    private val context: Context

    private val entries = ArrayDeque<LogEntry>(BUFFER_SIZE + 1)

    private var onLogListener: OnLogListener? = null

    init {
        if (!HAS_TIMBER) {
            throw RuntimeException("Timber dependency is not found")
        }
        this.context = context.applicationContext
    }

    fun tree(): Timber.Tree {
        return object : Timber.DebugTree() {
            override fun log(priority: Int, tag: String?, message: String, t: Throwable?) {
                addEntry(LogEntry(priority, tag!!, message,
                        logDatePattern.format(Calendar.getInstance().time)))
            }
        }
    }

    fun setOnLogListener(onLogListener: OnLogListener?) {
        this.onLogListener = onLogListener
    }

    @Synchronized private fun addEntry(entry: LogEntry) {
        entries.addLast(entry)

        if (entries.size > BUFFER_SIZE) {
            entries.removeFirst()
        }

        onLog(entry)
    }

    fun bufferedLogs(): List<LogEntry> {
        return ArrayList(entries)
    }

    /**
     * Save the current logs to disk.
     */
    fun save(listener: OnSaveLogListener) {
        val dir = logDir

        if (dir == null) {
            listener.onError(context.getString(R.string.all_save_logs_error))
            return
        }

        var fileWriter: FileWriter? = null

        try {
            val output = File(dir, logFileName)
            fileWriter = FileWriter(output, true)

            val entries = bufferedLogs()
            for (entry in entries) {
                fileWriter.write(entry.prettyPrint() + "\n")
            }

            listener.onSave(output)

        } catch (exception: IOException) {
            listener.onError(exception.message ?: "")
            Timber.w(exception)
        } finally {
            if (fileWriter != null) {
                try {
                    fileWriter.close()
                } catch (exception: IOException) {
                    listener.onError(exception.message ?: "")
                    Timber.w(exception)
                }

            }
        }
    }

    fun cleanUp() {
        val dir = logDir
        if (dir != null) {
            val files = dir.listFiles()
            for (file in files) {
                if (file.name.endsWith(LOG_FILE_END)) {
                    file.delete()
                }
            }
        }
    }

    private val logDir: File?
        get() {
            val file = File(context.filesDir, "logs")
            if (!file.exists()) {
                file.mkdirs()
            }
            return file
        }

    private fun onLog(entry: LogEntry) {
        if (onLogListener != null) {
            onLogListener!!.onLog(entry)
        }
    }

    private val logFileName: String
        get() {
            val pattern = "%s%s"
            val currentDate = filenameDate.format(Calendar.getInstance().time)

            return String.format(pattern, currentDate, LOG_FILE_END)
        }

    interface OnSaveLogListener {
        fun onSave(file: File)

        fun onError(message: String)
    }


    interface OnLogListener {
        fun onLog(logEntry: LogEntry)
    }

    companion object {

        private val HAS_TIMBER: Boolean

        init {
            var hasDependency: Boolean

            try {
                Class.forName("timber.log.Timber")
                hasDependency = true
            } catch (exception: ClassNotFoundException) {
                hasDependency = false
            }

            HAS_TIMBER = hasDependency
        }

        private val BUFFER_SIZE = 200

        private val LOG_FILE_END = ".log"

        @SuppressLint("StaticFieldLeak") private var sInstance: LumberYard? = null

        fun getInstance(context: Context): LumberYard {
            if (sInstance == null) {
                sInstance = LumberYard(context)
            }

            return sInstance!!
        }
    }
}